File: gemfilter/typescript/src/processors.ts
/**
 * Processor interface for SandFilter
 */
export interface Processor {
  process(match: string, ruleName: string): string;
}

/**
 * Replace sensitive text with a replacement string
 */
export class ReplaceProcessor implements Processor {
  constructor(private replacement: string = '[REDACTED]') {}

  process(_match: string, _ruleName: string): string {
    return this.replacement;
  }
}

/**
 * Replace with rule name in brackets
 */
export class RuleNameReplaceProcessor implements Processor {
  process(_match: string, ruleName: string): string {
    return `[${ruleName.toUpperCase()}]`;
  }
}

/**
 * Partially mask sensitive text
 */
export class PartialMaskProcessor implements Processor {
  constructor(
    private maskChar: string = '*',
    private preservePrefix: number = 1,
    private preserveSuffix: number = 1
  ) {}

  process(match: string, _ruleName: string): string {
    if (match.length <= this.preservePrefix + this.preserveSuffix) {
      return this.maskChar.repeat(match.length);
    }

    const prefix = match.slice(0, this.preservePrefix);
    const suffix = this.preserveSuffix > 0 ? match.slice(-this.preserveSuffix) : '';
    const middleLength = match.length - this.preservePrefix - this.preserveSuffix;

    // For email, try to preserve domain structure
    if (match.includes('@')) {
      const parts = match.split('@');
      if (parts.length === 2) {
        const [local, domain] = parts;
        const maskedLocal =
          local.length > 1
            ? local[0] + this.maskChar.repeat(local.length - 1)
            : this.maskChar;
        return `${maskedLocal}@${domain}`;
      }
    }

    return prefix + this.maskChar.repeat(middleLength) + suffix;
  }
}

/**
 * Completely remove sensitive text
 */
export class DeleteProcessor implements Processor {
  process(_match: string, _ruleName: string): string {
    return '';
  }
}

/**
 * Hash sensitive text using SHA256
 */
export class HashProcessor implements Processor {
  private truncate: boolean;

  constructor(truncate: boolean = true) {
    this.truncate = truncate;
  }

  async process(match: string, _ruleName: string): Promise<string> {
    // Simple hash implementation without external dependencies
    let hash = 0;
    for (let i = 0; i < match.length; i++) {
      const char = match.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32bit integer
    }
    const hashHex = Math.abs(hash).toString(16).padStart(8, '0');
    return this.truncate ? `[HASH:${hashHex}]` : `[HASH:${hashHex}]`;
  }
}

/**
 * Custom processor function
 */
export class CustomProcessor implements Processor {
  constructor(private func: (match: string, ruleName: string) => string) {}

  process(match: string, ruleName: string): string {
    return this.func(match, ruleName);
  }
}

/**
 * Processor factory
 */
export class Processors {
  static replace(replacement: string = '[REDACTED]'): ReplaceProcessor {
    return new ReplaceProcessor(replacement);
  }

  static ruleName(): RuleNameReplaceProcessor {
    return new RuleNameReplaceProcessor();
  }

  static partialMask(
    maskChar: string = '*',
    preservePrefix: number = 1,
    preserveSuffix: number = 1
  ): PartialMaskProcessor {
    return new PartialMaskProcessor(maskChar, preservePrefix, preserveSuffix);
  }

  static delete(): DeleteProcessor {
    return new DeleteProcessor();
  }

  static hash(truncate: boolean = true): HashProcessor {
    return new HashProcessor(truncate);
  }

  static custom(func: (match: string, ruleName: string) => string): CustomProcessor {
    return new CustomProcessor(func);
  }

  // Common presets
  static get EMAIL(): Processor {
    return Processors.ruleName();
  }

  static get PHONE(): Processor {
    return Processors.partialMask('*', 3, 4);
  }

  static get API_KEY(): Processor {
    return Processors.replace('[API_KEY]');
  }

  static get PASSWORD(): Processor {
    return Processors.replace('[PASSWORD]');
  }

  static get CREDIT_CARD(): Processor {
    return Processors.partialMask('*', 4, 4);
  }
}

File: sandfilter/core/processors.py
"""
Processor engine for SandFilter.
Defines how detected sensitive information is processed.
"""

import hashlib
import base64
from typing import Callable, Optional, Protocol
from dataclasses import dataclass


class Processor(Protocol):
    """Protocol for text processors."""

    def process(self, match: str, rule_name: str) -> str:
        """Process a matched string.

        Args:
            match: The matched sensitive text
            rule_name: The name of the rule that matched

        Returns:
            The processed (masked/replaced/etc) text
        """
        ...


@dataclass
class ProcessorConfig:
    """Configuration for processors."""
    replacement: str = "[REDACTED]"
    mask_char: str = "*"
    preserve_prefix: int = 1
    preserve_suffix: int = 1
    encryption_key: Optional[str] = None


class ReplaceProcessor:
    """Replace sensitive text with a replacement string.

    Example: "test@example.com" -> "[EMAIL]"
    """

    def __init__(self, replacement: str = "[REDACTED]"):
        self.replacement = replacement

    def process(self, match: str, rule_name: str) -> str:
        return self.replacement


class RuleNameReplaceProcessor:
    """Replace sensitive text with rule name in brackets.

    Example: "test@example.com" -> "[EMAIL]"
    """

    def process(self, match: str, rule_name: str) -> str:
        return f"[{rule_name.upper()}]"


class PartialMaskProcessor:
    """Partially mask sensitive text, preserving some characters.

    Example: "test@example.com" -> "te**@******.com"
    """

    def __init__(
        self,
        mask_char: str = "*",
        preserve_prefix: int = 1,
        preserve_suffix: int = 1,
    ):
        self.mask_char = mask_char
        self.preserve_prefix = preserve_prefix
        self.preserve_suffix = preserve_suffix

    def process(self, match: str, rule_name: str) -> str:
        if len(match) <= self.preserve_prefix + self.preserve_suffix:
            return self.mask_char * len(match)

        prefix = match[: self.preserve_prefix]
        suffix = match[-self.preserve_suffix :] if self.preserve_suffix > 0 else ""
        middle_length = len(match) - self.preserve_prefix - self.preserve_suffix

        # For email, try to preserve domain structure
        if "@" in match:
            parts = match.split("@")
            if len(parts) == 2:
                local, domain = parts
                masked_local = (
                    local[:1] + self.mask_char * (len(local) - 1)
                    if len(local) > 1
                    else self.mask_char
                )
                return f"{masked_local}@{domain}"

        return prefix + self.mask_char * middle_length + suffix


class DeleteProcessor:
    """Completely remove sensitive text."""

    def process(self, match: str, rule_name: str) -> str:
        return ""


class EncryptProcessor:
    """Encrypt sensitive text using AES (base64 encoded for display).

    Requires a secret key to be configured.
    """

    def __init__(self, secret_key: str):
        self.secret_key = secret_key.encode()[:32]  # Use first 32 bytes

    def process(self, match: str, rule_name: str) -> str:
        # Simple XOR encryption + base64 (for demo purposes)
        # In production, use proper AES from cryptography library
        key_bytes = self.secret_key
        encrypted = bytes(a ^ b for a, b in zip(match.encode(), (key_bytes * (len(match) // len(key_bytes) + 1))[:len(match)]))
        return f"[ENC:{base64.b64encode(encrypted).decode()}]"


class HashProcessor:
    """Hash sensitive text using SHA256 (one-way, irreversible)."""

    def __init__(self, truncate: bool = True):
        self.truncate = truncate

    def process(self, match: str, rule_name: str) -> str:
        hash_value = hashlib.sha256(match.encode()).hexdigest()
        if self.truncate:
            return f"[HASH:{hash_value[:16]}]"
        return f"[HASH:{hash_value}]"


class CustomProcessor:
    """User-defined custom processor function."""

    def __init__(self, func: Callable[[str, str], str]):
        self.func = func

    def process(self, match: str, rule_name: str) -> str:
        return self.func(match, rule_name)


class FixedValueProcessor:
    """Replace with a fixed value (useful for consistent masking)."""

    def __init__(self, value: str):
        self.value = value

    def process(self, match: str, rule_name: str) -> str:
        return self.value


# Built-in processor factory
class Processors:
    """Factory class for creating built-in processors."""

    @staticmethod
    def replace(replacement: str = "[REDACTED]") -> ReplaceProcessor:
        return ReplaceProcessor(replacement)

    @staticmethod
    def rule_name() -> RuleNameReplaceProcessor:
        return RuleNameReplaceProcessor()

    @staticmethod
    def partial_mask(
        mask_char: str = "*",
        preserve_prefix: int = 1,
        preserve_suffix: int = 1,
    ) -> PartialMaskProcessor:
        return PartialMaskProcessor(mask_char, preserve_prefix, preserve_suffix)

    @staticmethod
    def delete() -> DeleteProcessor:
        return DeleteProcessor()

    @staticmethod
    def encrypt(secret_key: str) -> EncryptProcessor:
        return EncryptProcessor(secret_key)

    @staticmethod
    def hash(truncate: bool = True) -> HashProcessor:
        return HashProcessor(truncate)

    @staticmethod
    def custom(func: Callable[[str, str], str]) -> CustomProcessor:
        return CustomProcessor(func)

    @staticmethod
    def fixed(value: str) -> FixedValueProcessor:
        return FixedValueProcessor(value)

    # Common presets
    EMAIL = rule_name()
    PHONE = partial_mask(preserve_prefix=3, preserve_suffix=4)
    API_KEY = replace("[API_KEY]")
    PASSWORD = replace("[PASSWORD]")
    CREDIT_CARD = partial_mask(preserve_prefix=4, preserve_suffix=4)
